File: lib/to_mdx/parsers/open_document/draw/path.rb
# frozen_string_literal: true

require_relative "../base"

require_relative "../../../detectors/code_detector"
require_relative "../mixins/positionable_element"
require_relative "../mixins/svg"
require_relative "../mixins/has_formula"

module ToMdx
  module Parsers
    module Draw
      class Path < Base
        include Mixins::HasFormula
        include Mixins::PositionableElement
        include Mixins::Svg

        def initialize(element = nil, context = {})
          super(element, context)
          @draw_style_name = context[:styles].use(element&.attributes&.[]("draw:style-name"))
        end

        def empty?
          false
        end

        def to_mdx
          wrap_with_svg_tag("<path d=\"#{svg_path}\" />", @draw_style_name)
        end

        def to_s
          contentful_children.map(&:to_s).compact.join
        end

        register_for "draw:path"
      end
    end
  end
end

File: src/parsers/open-document/draw/path.ts
import type { Style } from "../../../styles.ts";
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";
import { FormulaEvaluator } from "../mixins/formula_evaluator.ts";

export class Path extends BaseElement {
  private drawStyleName: Style;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.drawStyleName = this.context("styles").use(this.attr("draw:style-name"));
  }

  empty(): boolean {
    return false;
  }

  toMdx(): string {
    const svgPath = this.hasFormula.svgPath();
    if (!svgPath) {
      return "";
    }

    const classAttr = this.drawStyleName.empty() ? "" : ` className="${this.drawStyleName}"`;
    return `<path${classAttr} d="${svgPath}" />`;
  }

  toString(): string {
    return this.contentfulChildren()
      .map((child) => child.toString())
      .filter(Boolean)
      .join("");
  }

  private get hasFormula(): FormulaEvaluator {
    return new FormulaEvaluator(this);
  }

  static {
    BaseElement.registerFor.call(this, "draw:path");
  }
}
